File: src/main/java/net/gtaun/shoebill/SampObjectStoreImpl.java
/**
 * Copyright (C) 2011-2014 MK124
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.gtaun.shoebill;

import net.gtaun.shoebill.data.SpawnInfo;
import net.gtaun.shoebill.event.dialog.DialogResponseEvent;
import net.gtaun.shoebill.event.object.ObjectMovedEvent;
import net.gtaun.shoebill.event.object.PlayerObjectMovedEvent;
import net.gtaun.shoebill.event.player.PlayerDisconnectEvent;
import net.gtaun.shoebill.event.player.PlayerUpdateEvent;
import net.gtaun.shoebill.event.vehicle.VehicleModEvent;
import net.gtaun.shoebill.event.vehicle.VehicleUpdateDamageEvent;
import net.gtaun.shoebill.object.*;
import net.gtaun.shoebill.object.Timer;
import net.gtaun.shoebill.object.impl.*;
import net.gtaun.util.event.EventManager;
import net.gtaun.util.event.HandlerPriority;

import java.lang.ref.Reference;
import java.lang.ref.WeakReference;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;

/**
 * @author MK124 & 123marvin123
 */
public class SampObjectStoreImpl implements SampObjectStore
{
    public static final int MAX_PLAYERS = 1000;
    public static final int MAX_VEHICLES = 2000;
    public static final int MAX_OBJECTS = 1000;
    public static final int MAX_ZONES = 1024;
    public static final int MAX_TEXT_DRAWS = 2048;
    public static final int MAX_PLAYER_TEXT_DRAWS = 256;
    public static final int MAX_MENUS = 128;
    public static final int MAX_GLOBAL_LABELS = 1024;
    public static final int MAX_PLAYER_LABELS = 1024;
    public static final int MAX_PICKUPS = 4096;
    public static final int MAX_CLASSES = 300;
    public static final int MAX_ACTORS = 1000;

    private static <T> Collection<T> getNotNullInstances(T[] items) {
        Collection<T> list = new ArrayList<>();
        if (items == null) return list;

        for (T item : items) {
            if (item != null) list.add(item);
        }
        return list;
    }

    private static <T> void clearUnusedReferences(Collection<Reference<T>> collection) {
        Iterator<Reference<T>> iterator = collection.iterator();
        while (iterator.hasNext()) {
            if (iterator.next().get() == null) iterator.remove();
        }
    }


    protected final EventManager eventManagerNode;

    private ServerImpl server;
    private WorldImpl world;
    private PlayerImpl[] players = new PlayerImpl[MAX_PLAYERS];
    private VehicleImpl[] vehicles = new VehicleImpl[MAX_VEHICLES];
    private SampObjectImpl[] objects = new SampObjectImpl[MAX_OBJECTS];
    private PlayerObjectImpl[][] playerObjectsArray = new PlayerObjectImpl[MAX_PLAYERS][];
    private PickupImpl[] pickups = new PickupImpl[MAX_PICKUPS];
    private LabelImpl[] labels = new LabelImpl[MAX_GLOBAL_LABELS];
    private PlayerLabelImpl[][] playerLabelsArray = new PlayerLabelImpl[MAX_PLAYERS][];
    private TextdrawImpl[] textdraws = new TextdrawImpl[MAX_TEXT_DRAWS];
    private PlayerTextdrawImpl[][] playerTextdrawsArray = new PlayerTextdrawImpl[MAX_PLAYERS][];
    private ZoneImpl[] zones = new ZoneImpl[MAX_ZONES];
    private MenuImpl[] menus = new MenuImpl[MAX_MENUS];
    private ActorImpl[] actors = new ActorImpl[MAX_ACTORS];
    private SpawnInfo[] playerClasses = new SpawnInfo[MAX_CLASSES];
    private Collection<Reference<TimerImpl>> timers = new ConcurrentLinkedQueue<>();
    private Map<Integer, Reference<DialogIdImpl>> dialogs = new ConcurrentHashMap<>();
    private List<Reference<PickupImpl>> staticPickups = new ArrayList<>();


    SampObjectStoreImpl(EventManager rootEventManager)
    {
        eventManagerNode = rootEventManager.createChildNode();
        setupObjectEventHandler();
    }

    private void setupObjectEventHandler() {
        eventManagerNode.registerHandler(PlayerUpdateEvent.class, HandlerPriority.MONITOR, (e) ->
        {
            Player primitive = e.getPlayer().getPrimitive();
            if (!(primitive instanceof PlayerImpl)) return;

            PlayerImpl player = (PlayerImpl) primitive;
            player.onPlayerUpdate();
        });

        eventManagerNode.registerHandler(PlayerDisconnectEvent.class, HandlerPriority.BOTTOM, (e) ->
        {
            Player primitive = e.getPlayer().getPrimitive();
            if (!(primitive instanceof PlayerImpl)) return;

            PlayerImpl player = (PlayerImpl) primitive;
            player.onPlayerDisconnect();
        });

        eventManagerNode.registerHandler(DialogResponseEvent.class, HandlerPriority.MONITOR, (e) ->
        {
            Player primitive = e.getPlayer().getPrimitive();
            if (!(primitive instanceof PlayerImpl)) return;

            PlayerImpl player = (PlayerImpl) primitive;
            player.onDialogResponse();
        });

        eventManagerNode.registerHandler(PlayerObjectMovedEvent.class, HandlerPriority.MONITOR, (e) ->
        {
            PlayerObject primitive = (PlayerObject) e.getObject().getPrimitive();
            if (!(primitive instanceof PlayerObjectImpl)) return;

            PlayerObjectImpl object = (PlayerObjectImpl) primitive;
            object.onPlayerObjectMoved();
        });

        eventManagerNode.registerHandler(ObjectMovedEvent.class, HandlerPriority.MONITOR, (e) ->
        {
            SampObject primitive = e.getObject().getPrimitive();
            if (!(primitive instanceof SampObjectImpl)) return;

            SampObjectImpl object = (SampObjectImpl) primitive;
            object.onObjectMoved();
        });

        eventManagerNode.registerHandler(VehicleModEvent.class, HandlerPriority.MONITOR, (e) ->
        {
            Vehicle primitive = e.getVehicle().getPrimitive();
            if (!(primitive instanceof VehicleImpl)) return;

            VehicleImpl vehicle = (VehicleImpl) primitive;
            vehicle.onVehicleMod();
        });

        eventManagerNode.registerHandler(VehicleUpdateDamageEvent.class, HandlerPriority.MONITOR, (e) ->
        {
            Vehicle primitive = e.getVehicle().getPrimitive();
            if (!(primitive instanceof VehicleImpl)) return;

            VehicleImpl vehicle = (VehicleImpl) primitive;
            vehicle.onVehicleUpdateDamage();
        });
    }

    @Override
    public Server getServer() {
        return server;
    }

    public void setServer(ServerImpl server) {
        this.server = server;
    }

    @Override
    public World getWorld() {
        return world;
    }

    public void setWorld(WorldImpl world) {
        this.world = world;
    }

    @Override
    public Player getPlayer(int id) {
        if (id < 0 || id >= MAX_PLAYERS) return null;
        return players[id];
    }

    @Override
    public Player getPlayer(String name) {
        if (name == null) return null;
        name = name.trim();

        for (Player player : players) {
            if (player == null) continue;
            if (player.getName().equalsIgnoreCase(name)) return player;
        }

        return null;
    }

    @Override
    public Vehicle getVehicle(int id) {
        if (id < 0 || id >= MAX_VEHICLES) return null;
        return vehicles[id];
    }

    @Override
    public SampObject getObject(int id) {
        if (id < 0 || id >= MAX_OBJECTS) return null;
        return objects[id];
    }

    @Override
    public PlayerObject getPlayerObject(Player player, int id) {
        if (player == null) return null;

        int playerId = player.getId();
        if (playerId < 0 || playerId >= MAX_PLAYERS) return null;

        if (id < 0 || id >= MAX_OBJECTS) return null;

        PlayerObject[] playerObjects = playerObjectsArray[playerId];
        if (playerObjects == null) return null;

        return playerObjects[id];
    }

    @Override
    public Pickup getPickup(int id) {
        if (id < 0 || id >= MAX_PICKUPS) return null;
        return pickups[id];
    }

    @Override
    public Label getLabel(int id) {
        if (id < 0 || id >= MAX_GLOBAL_LABELS) return null;
        return labels[id];
    }

    @Override
    public PlayerLabel getPlayerLabel(Player player, int id) {
        if (player == null) return null;

        int playerId = player.getId();
        if (playerId < 0 || playerId >= MAX_PLAYERS) return null;

        if (id < 0 || id >= MAX_PLAYER_LABELS) return null;

        PlayerLabel[] playerLabels = playerLabelsArray[playerId];
        if (playerLabels == null) return null;

        return playerLabels[id];
    }

    @Override
    public Textdraw getTextdraw(int id) {
        if (id < 0 || id >= MAX_TEXT_DRAWS) return null;
        return textdraws[id];
    }

    @Override
    public PlayerTextdraw getPlayerTextdraw(Player player, int id) {
        if (player == null) return null;

        int playerId = player.getId();
        if (playerId < 0 || playerId >= MAX_PLAYERS) return null;

        if (id < 0 || id >= MAX_PLAYER_TEXT_DRAWS) return null;

        PlayerTextdraw[] textdraws = playerTextdrawsArray[playerId];
        if (textdraws == null) return null;

        return textdraws[id];
    }

    @Override
    public Zone getZone(int id) {
        if (id < 0 || id >= MAX_ZONES) return null;
        return zones[id];
    }

    @Override
    public Menu getMenu(int id) {
        if (id < 0 || id >= MAX_MENUS) return null;
        return menus[id];
    }

    @Override
    public DialogId getDialog(int id) {
        Reference<DialogIdImpl> ref = dialogs.get(id);
        return ref != null ? ref.get() : null;
    }

    @Override
    public Collection<Player> getPlayers() {
        return getNotNullInstances(players);
    }

    @Override
    public Collection<Player> getHumanPlayers() {
        Collection<Player> list = new ArrayList<>();
        if (players == null) return list;

        for (Player item : players) {
            if (item != null && !item.isNpc()) list.add(item);
        }
        return list;
    }

    @Override
    public Collection<Player> getNpcPlayers() {
        Collection<Player> list = new ArrayList<>();
        if (players == null) return list;

        for (Player item : players) {
            if (item != null && item.isNpc()) list.add(item);
        }
        return list;
    }

    @Override
    public Collection<Vehicle> getVehicles() {
        return getNotNullInstances(vehicles);
    }

    @Override
    public Collection<SampObject> getObjects() {
        return getNotNullInstances(objects);
    }

    @Override
    public Collection<PlayerObject> getPlayerObjects(Player player) {
        if (player == null) return null;

        int playerId = player.getId();
        if (playerId < 0 || playerId >= MAX_PLAYERS) return new ArrayList<>(0);

        PlayerObject[] playerObjects = playerObjectsArray[playerId];
        return getNotNullInstances(playerObjects);
    }

    @Override
    public Collection<Pickup> getPickups() {
        return getNotNullInstances(pickups);
    }

    @Override
    public Collection<Label> getLabels() {
        return getNotNullInstances(labels);
    }

    @Override
    public Collection<PlayerLabel> getPlayerLabels(Player player) {
        if (player == null) return null;

        int playerId = player.getId();
        if (playerId < 0 || playerId >= MAX_PLAYERS) return new ArrayList<>(0);

        PlayerLabel[] playerLabels = playerLabelsArray[playerId];
        return getNotNullInstances(playerLabels);
    }

    @Override
    public Collection<Textdraw> getTextdraws() {
        return getNotNullInstances(textdraws);
    }

    @Override
    public Collection<PlayerTextdraw> getPlayerTextdraws(Player player) {
        if (player == null) return null;

        int playerId = player.getId();
        if (playerId < 0 || playerId >= MAX_PLAYERS) return new ArrayList<>(0);

        PlayerTextdraw[] textdraws = playerTextdrawsArray[playerId];
        return getNotNullInstances(textdraws);
    }

    @Override
    public Collection<Zone> getZones() {
        return getNotNullInstances(zones);
    }

    @Override
    public Collection<Menu> getMenus() {
        return getNotNullInstances(menus);
    }

    @Override
    public Collection<DialogId> getDialogIds() {
        Collection<DialogId> items = new ArrayList<>();
        for (Reference<DialogIdImpl> reference : dialogs.values()) {
            DialogId dialog = reference.get();
            if (dialog != null) items.add(dialog);
        }

        return items;
    }

    @Override
    public Collection<SpawnInfo> getPlayerClasses() {
        Collection<SpawnInfo> items = new ArrayList<>();
        Collections.addAll(items, playerClasses);
        return items;
    }

    @Override
    public Collection<Actor> getActors() {
        return new ArrayList<>(Arrays.asList(actors));
    }

    @Override
    public Actor getActor(int id) {
        if (id < 0 || id > MAX_ACTORS) return null;
        return actors[id];
    }

    @Override
    public int getVehiclePoolSize() {
        return SampNativeFunction.getVehiclePoolSize();
    }

    @Override
    public int getPlayerPoolSize() {
        return SampNativeFunction.getPlayerPoolSize();
    }

    @Override
    public int getActorPoolSize() {
        return SampNativeFunction.getActorPoolSize();
    }

    public Collection<Timer> getTimers() {
        Collection<Timer> items = new ArrayList<>();
        Collection<Reference<TimerImpl>> unusedItems = new ArrayList<>();
        for (Reference<TimerImpl> reference : timers) {
            Timer timer = reference.get();
            if (timer != null) items.add(timer);
            else unusedItems.add(reference);
        }

        timers.removeAll(unusedItems);
        return items;
    }

    public void setPlayer(int id, PlayerImpl player) {
        if (player == null && players[id] != null) {
            removePlayer(id);
            if (player == null) return;
        }

        players[id] = player;

        playerObjectsArray[id] = new PlayerObjectImpl[MAX_OBJECTS];
        playerLabelsArray[id] = new PlayerLabelImpl[MAX_PLAYER_LABELS];
        playerTextdrawsArray[id] = new PlayerTextdrawImpl[MAX_PLAYER_TEXT_DRAWS];
    }

    private void removePlayer(int id) {
        for (PlayerLabel playerLabel : playerLabelsArray[id]) {
            if (playerLabel != null) playerLabel.destroy();
        }

        for (PlayerObject playerObject : playerObjectsArray[id]) {
            if (playerObject != null) playerObject.destroy();
        }

        playerLabelsArray[id] = null;
        playerObjectsArray[id] = null;
        playerTextdrawsArray[id] = null;
        players[id] = null;
    }

    public void setVehicle(int id, VehicleImpl vehicle) {
        vehicles[id] = vehicle;
    }

    public void setObject(int id, SampObjectImpl object) {
        objects[id] = object;
    }

    public void setPlayerObject(Player player, int id, PlayerObject object) {
        if (!player.isOnline()) return;
        PlayerObject[] playerObjects = playerObjectsArray[player.getId()];
        playerObjects[id] = object;
    }

    public void setPickup(int id, PickupImpl pickup) {
        pickups[id] = pickup;
    }

    public void setLabel(int id, LabelImpl label) {
        labels[id] = label;
    }

    public void setPlayerLabel(Player player, int id, PlayerLabelImpl label) {
        if (!player.isOnline() || label == null) return;
        PlayerLabel[] playerLabels = playerLabelsArray[player.getId()];
        playerLabels[id] = label;
    }

    public void setTextdraw(int id, TextdrawImpl textdraw) {
        textdraws[id] = textdraw;
    }

    public void setPlayerTextdraw(Player player, int id, PlayerTextdraw textdraw) {
        if (!player.isOnline()) return;
        PlayerTextdraw[] textdraws = playerTextdrawsArray[player.getId()];
        textdraws[id] = textdraw;
    }

    public void setZone(int id, ZoneImpl zone) {
        zones[id] = zone;
    }

    public void setMenu(int id, MenuImpl menu) {
        menus[id] = menu;
    }

    public void setPlayerClass(int id, SpawnInfo playerClass) { playerClasses[id] = playerClass; }

    public void setActor(int id, ActorImpl actor) {
        actors[id] = actor;
    }

    public void putTimer(TimerImpl timer) {
        clearUnusedReferences(timers);
        timers.add(new WeakReference<>(timer));
    }

    public void removeTimer(Timer timer) {
        for (Iterator<Reference<TimerImpl>> iterator = timers.iterator(); iterator.hasNext(); ) {
            Reference<TimerImpl> ref = iterator.next();
            Timer t = ref.get();
            if (t == null || t == timer) iterator.remove();
        }
    }

    public void putDialog(int id, DialogIdImpl dialog) {
        clearUnusedReferences(dialogs.values());
        dialogs.put(id, new WeakReference<>(dialog));
    }

    public void removeDialog(DialogId dialog) {
        dialogs.remove(dialog.getId());
    }

    public void addStaticPickup(PickupImpl pickup) {
        if (pickup != null && pickup.isStatic()) {
            clearUnusedReferences(staticPickups);
            staticPickups.add(new WeakReference<>(pickup));
        }
    }

    public void removeStaticPickup(Pickup pickup) {
        for (Iterator<Reference<PickupImpl>> iterator = staticPickups.iterator(); iterator.hasNext(); ) {
            Reference<PickupImpl> ref = iterator.next();
            Pickup p = ref.get();
            if (p == null || p == pickup) iterator.remove();
        }
    }

    public List<PickupImpl> getStaticPickups() {
        List<PickupImpl> items = new ArrayList<>();
        Collection<Reference<PickupImpl>> unusedItems = new ArrayList<>();
        for (Reference<PickupImpl> reference : staticPickups) {
            PickupImpl pickup = reference.get();
            if (pickup != null) items.add(pickup);
            else unusedItems.add(reference);
        }

        staticPickups.removeAll(unusedItems);
        return items;
    }

    public Collection<TimerImpl> getTimerImpls() {
        Collection<TimerImpl> items = new ArrayList<>();
        Collection<Reference<TimerImpl>> unusedItems = new ArrayList<>();
        for (Reference<TimerImpl> reference : timers) {
            TimerImpl timer = reference.get();
            if (timer != null) items.add(timer);
            else unusedItems.add(reference);
        }

        timers.removeAll(unusedItems);
        return items;
    }
}

File: src/main/java/net/gtaun/shoebill/SampObjectStoreImpl.kt
/**
 * Copyright (C) 2011-2014 MK124

 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at

 * http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.gtaun.shoebill

import net.gtaun.shoebill.data.SpawnInfo
import net.gtaun.shoebill.entities.*
import net.gtaun.shoebill.entities.impl.*
import net.gtaun.shoebill.event.dialog.DialogResponseEvent
import net.gtaun.shoebill.event.player.PlayerDisconnectEvent
import net.gtaun.shoebill.event.player.PlayerUpdateEvent
import net.gtaun.shoebill.event.sampobject.ObjectMovedEvent
import net.gtaun.shoebill.event.sampobject.PlayerObjectMovedEvent
import net.gtaun.shoebill.event.vehicle.VehicleModEvent
import net.gtaun.shoebill.event.vehicle.VehicleUpdateDamageEvent
import net.gtaun.util.event.EventManager
import net.gtaun.util.event.HandlerPriority
import java.lang.ref.Reference
import java.lang.ref.WeakReference
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedQueue

/**
 * @author MK124 & 123marvin123
 */
open class SampObjectStoreImpl internal constructor(rootEventManager: EventManager) : SampObjectStore {

    internal val eventManagerNode: EventManager = rootEventManager.createChildNode()

    lateinit override var server: ServerImpl
    lateinit override var world: WorldImpl
    override val players: Array<Player?> = arrayOfNulls(MAX_PLAYERS)
    override val vehicles: Array<Vehicle?> = arrayOfNulls(MAX_VEHICLES)
    override val objects: Array<SampObject?> = arrayOfNulls(MAX_OBJECTS)
    private val playerObjectsArray: Array<Array<PlayerObject?>?> = arrayOfNulls(MAX_PLAYERS)
    override val playerClasses: Array<SpawnInfo?> = arrayOfNulls(MAX_CLASSES)
    override val pickups: Array<Pickup?> = arrayOfNulls(MAX_PICKUPS)
    override val labels: Array<Label?> = arrayOfNulls(MAX_GLOBAL_LABELS)
    private val playerLabelsArray: Array<Array<PlayerLabel?>?> = arrayOfNulls(MAX_PLAYERS)
    override val textdraws: Array<Textdraw?> = arrayOfNulls(MAX_TEXT_DRAWS)
    private val playerTextdrawsArray: Array<Array<PlayerTextdraw?>?> = arrayOfNulls(MAX_PLAYERS)
    override val menus: Array<Menu?> = arrayOfNulls(MAX_MENUS)
    override val zones: Array<Zone?> = arrayOfNulls(MAX_ZONES)
    override val actors: Array<Actor?> = arrayOfNulls(MAX_ACTORS)
    private val timers = ConcurrentLinkedQueue<Reference<TimerImpl>>()
    private val dialogs = ConcurrentHashMap<Int, Reference<DialogIdImpl>>()
    private val staticPickups: MutableList<Reference<PickupImpl>> = mutableListOf()


    init {
        setupObjectEventHandler()
    }

    private fun setupObjectEventHandler() {
        eventManagerNode.registerHandler(PlayerUpdateEvent::class, {
            val primitive = it.player.primitive as? PlayerImpl ?: return@registerHandler
            primitive.onPlayerUpdate()
        }, HandlerPriority.MONITOR)

        eventManagerNode.registerHandler(PlayerDisconnectEvent::class, {
            val primitive = it.player.primitive as? PlayerImpl ?: return@registerHandler
            primitive.onPlayerDisconnect()
        }, HandlerPriority.BOTTOM)

        eventManagerNode.registerHandler(DialogResponseEvent::class, {
            val primitive = it.player.primitive as? PlayerImpl ?: return@registerHandler
            primitive.onDialogResponse()
        }, HandlerPriority.MONITOR)

        eventManagerNode.registerHandler(PlayerObjectMovedEvent::class, {
            val primitive = it.`object`.primitive as? PlayerObjectImpl ?: return@registerHandler
            primitive.onPlayerObjectMoved()
        }, HandlerPriority.MONITOR)

        eventManagerNode.registerHandler(ObjectMovedEvent::class, {
            val primitive = it.`object`.primitive as? SampObjectImpl ?: return@registerHandler
            primitive.onObjectMoved()
        }, HandlerPriority.MONITOR)

        eventManagerNode.registerHandler(VehicleModEvent::class, {
            val primitive = it.vehicle.primitive as? VehicleImpl ?: return@registerHandler
            primitive.onVehicleMod()
        }, HandlerPriority.MONITOR)

        eventManagerNode.registerHandler(VehicleUpdateDamageEvent::class, {
            val primitive = it.vehicle.primitive as? VehicleImpl ?: return@registerHandler
            primitive.onVehicleUpdateDamage()
        }, HandlerPriority.MONITOR)
    }

    override fun getPlayer(id: Int): Player? = players.getOrNull(id)

    override fun getVehicle(id: Int): Vehicle? = vehicles.getOrNull(id)

    override fun getObject(id: Int): SampObject? = objects.getOrNull(id)

    override fun getPickup(id: Int): Pickup? = pickups.getOrNull(id)

    override fun getStaticPickup(id: Int): Pickup? = staticPickups.getOrNull(id)?.get()

    override fun getLabel(id: Int): Label? = labels.getOrNull(id)

    override fun getTextdraw(id: Int): Textdraw? = textdraws.getOrNull(id)

    override fun getZone(id: Int): Zone? = zones.getOrNull(id)

    override fun getMenu(id: Int): Menu? = menus.getOrNull(id)

    override fun getDialog(id: Int): DialogId? = dialogs[id]?.get()

    override val humanPlayers: Collection<Player>
        get() = players.filterNotNull().filter { !it.isNpc }

    override val npcPlayers: Collection<Player>
        get() = players.filterNotNull().filter { it.isNpc }

    override val dialogIds: Collection<DialogId>
        get() = dialogs.values.filter { it.get() != null }.map { it.get()!! }

    override fun getActor(id: Int): Actor? = actors.getOrNull(id)

    override val vehiclePoolSize: Int
        get() = SampNativeFunction.getVehiclePoolSize()

    override val playerPoolSize: Int
        get() = SampNativeFunction.getPlayerPoolSize()

    override val actorPoolSize: Int
        get() = SampNativeFunction.getActorPoolSize()

    fun getTimers(): Collection<Timer> {
        clearUnusedReferences(timers)
        return timers.map { it.get()!! }
    }

    internal fun setPlayer(id: Int, player: PlayerImpl) {
        players[id] = player

        playerObjectsArray[id] = arrayOfNulls(MAX_OBJECTS)
        playerLabelsArray[id] = arrayOfNulls(MAX_PLAYER_LABELS)
        playerTextdrawsArray[id] = arrayOfNulls(MAX_PLAYER_TEXT_DRAWS)
    }

    fun removePlayer(id: Int) {
        val playerLabels = playerLabelsArray.getOrNull(id)
        val playerObjects = playerObjectsArray.getOrNull(id)
        val playerTextdraws = playerTextdrawsArray.getOrNull(id)

        playerLabels?.filterNotNull()?.filter { !it.isDestroyed }?.forEach { it.destroy() }
        playerObjects?.filterNotNull()?.filter { !it.isDestroyed }?.forEach { it.destroy() }
        playerTextdraws?.filterNotNull()?.filter { !it.isDestroyed }?.forEach { it.destroy() }

        playerLabelsArray[id] = null
        playerObjectsArray[id] = null
        playerTextdrawsArray[id] = null

        players[id] = null
    }

    fun removeVehicle(id: Int) = vehicles.tryRemoveAtIndex(id)

    fun removePlayerObject(player: Player, id: Int) = playerObjectsArray[player.id]?.tryRemoveAtIndex(id)

    fun removeObject(id: Int) = objects.tryRemoveAtIndex(id)

    fun removePickup(id: Int) = pickups.tryRemoveAtIndex(id)

    fun removeLabel(id: Int) = labels.tryRemoveAtIndex(id)

    fun removePlayerLabel(player: Player, id: Int) = playerLabelsArray[player.id]?.tryRemoveAtIndex(id)

    fun removeTextdraw(id: Int) = textdraws.tryRemoveAtIndex(id)

    fun removePlayerTextdraw(player: Int, id: Int) = playerTextdrawsArray[player]?.tryRemoveAtIndex(id)

    fun removeZone(id: Int) = zones.tryRemoveAtIndex(id)

    fun removeMenu(id: Int) = menus.tryRemoveAtIndex(id)

    fun removeActor(id: Int) = actors.tryRemoveAtIndex(id)

    fun setVehicle(id: Int, vehicle: VehicleImpl) {
        vehicles[id] = vehicle
    }

    fun setObject(id: Int, `object`: SampObjectImpl) {
        objects[id] = `object`
    }

    fun setPlayerObject(player: Player, id: Int, `object`: PlayerObject) {
        if (!player.isOnline) return

        val playerObjects = playerObjectsArray[player.id] ?: return
        playerObjects[id] = `object`
    }

    fun setPickup(id: Int, pickup: PickupImpl) {
        pickups[id] = pickup
    }

    fun setLabel(id: Int, label: LabelImpl) {
        labels[id] = label
    }

    fun setPlayerLabel(player: Player, id: Int, label: PlayerLabelImpl) {
        if (!player.isOnline) return

        val playerLabels = playerLabelsArray[player.id] ?: return
        playerLabels[id] = label
    }

    fun setTextdraw(id: Int, textdraw: TextdrawImpl) {
        textdraws[id] = textdraw
    }

    fun setPlayerTextdraw(player: Player, id: Int, textdraw: PlayerTextdraw) {
        if (!player.isOnline) return

        val playerTextdraws = playerTextdrawsArray[player.id] ?: return
        playerTextdraws[id] = textdraw
    }

    fun setZone(id: Int, zone: ZoneImpl) {
        zones[id] = zone
    }

    fun setMenu(id: Int, menu: MenuImpl) {
        menus[id] = menu
    }

    fun setPlayerClass(id: Int, playerClass: SpawnInfo) {
        playerClasses[id] = playerClass
    }

    fun setActor(id: Int, actor: ActorImpl) {
        actors[id] = actor
    }

    internal fun putTimer(timer: TimerImpl) {
        clearUnusedReferences(timers)
        timers.add(WeakReference(timer))
    }

    internal fun removeTimer(timer: Timer) {
        clearUnusedReferences(timers)
        timers.apply { remove(find { it.get() === timer }) }
    }

    internal fun putDialog(id: Int, dialog: DialogIdImpl) {
        clearUnusedReferences(dialogs.values)
        dialogs.put(id, WeakReference(dialog))
    }

    internal fun removeDialog(dialog: DialogId) = dialogs.remove(dialog.id)

    override fun getPlayerTextdraws(player: Player): Collection<PlayerTextdraw> =
            playerTextdrawsArray[player.id]?.filterNotNull() ?: listOf()

    override fun getPlayerTextdraw(player: Player, id: Int): PlayerTextdraw? =
            getPlayerTextdraws(player).filter { it.id == id }.firstOrNull()

    override fun getPlayerObjects(player: Player): Collection<PlayerObject> =
            playerObjectsArray[player.id]?.filterNotNull() ?: listOf()

    override fun getPlayerObject(player: Player, id: Int): PlayerObject? =
            getPlayerObjects(player).filter { it.id == id }.firstOrNull()

    override fun getPlayerLabels(player: Player): Collection<PlayerLabel> =
            playerLabelsArray[player.id]?.filterNotNull() ?: listOf()

    override fun getPlayerLabel(player: Player, id: Int): PlayerLabel? =
            getPlayerLabels(player).filter { it.id == id }.firstOrNull()

    override fun getPlayer(name: String): Player? =
            players.filterNotNull().filter { it.name == name }.firstOrNull()

    fun addStaticPickup(pickup: PickupImpl) {
        if(!pickup.isStatic) return

        clearUnusedReferences(staticPickups)
        staticPickups.add(WeakReference(pickup))
    }

    fun removeStaticPickup(pickup: Pickup) {
        if(!pickup.isStatic) return

        clearUnusedReferences(staticPickups)
        staticPickups.apply { remove(find { it.get() === pickup }) }
    }

    internal fun getStaticPickups(): List<PickupImpl> {
        clearUnusedReferences(staticPickups)
        return staticPickups.map { it.get()!! }
    }

    internal val timerImpls: Collection<TimerImpl>
        get() {
            clearUnusedReferences(timers)
            return timers.map { it.get()!! }
        }


    companion object {

        @JvmField
        val MAX_PLAYERS = 1000

        @JvmStatic
        private val MAX_VEHICLES = 2000

        @JvmStatic
        private val MAX_OBJECTS = 1000

        @JvmStatic
        private val MAX_ZONES = 1024

        @JvmStatic
        private val MAX_TEXT_DRAWS = 2048

        @JvmStatic
        private val MAX_PLAYER_TEXT_DRAWS = 256

        @JvmStatic
        private val MAX_MENUS = 128

        @JvmStatic
        private val MAX_GLOBAL_LABELS = 1024

        @JvmStatic
        private val MAX_PLAYER_LABELS = 1024

        @JvmStatic
        private val MAX_PICKUPS = 4096

        @JvmStatic
        private val MAX_CLASSES = 320

        @JvmStatic
        private val MAX_ACTORS = 1000

        @JvmStatic
        private fun <T> clearUnusedReferences(collection: MutableCollection<Reference<T>>) {
            val iterator = collection.iterator()
            while (iterator.hasNext()) {
                if (iterator.next().get() == null) iterator.remove()
            }
        }
    }
}

fun <T> Array<T>.indexExists(id: Int): Boolean {
    return id >= 0 && id < this.size
}

fun <T> Array<T?>.tryRemoveAtIndex(index: Int): Boolean {
    if(!indexExists(index)) return false
    this[index] = null
    return true
}